In [2]:
import pandas as pd
%matplotlib inline
import matplotlib
import matplotlib.pyplot as plt
import numpy as np

In [3]:
#Reading users file:
u_cols = ['user_id', 'age', 'sex', 'occupation', 'zip_code']
users = pd.read_csv('ml-100k/u.user', sep='|', names=u_cols,encoding='latin-1')

In [4]:
#Reading ratings file:
r_cols = ['user_id', 'movie_id', 'rating', 'unix_timestamp']
ratings = pd.read_csv('ml-100k/u.data', sep='\t', names=r_cols,encoding='latin-1')

In [5]:
#Reading items file:
i_cols = ['movie id', 'movie title' ,'release date','video release date', 'IMDb URL', 'unknown', 'Action', 'Adventure',
'Animation', 'Children\'s', 'Comedy', 'Crime', 'Documentary', 'Drama', 'Fantasy',
'Film-Noir', 'Horror', 'Musical', 'Mystery', 'Romance', 'Sci-Fi', 'Thriller', 'War', 'Western']
items = pd.read_csv('ml-100k/u.item', sep='|', names=i_cols,
encoding='latin-1')

In [6]:
print(users.shape)
users.head()

(943, 5)


user_id  age sex  occupation zip_code
0        1   24   M  technician    85711
1        2   53   F       other    94043
2        3   23   M      writer    32067
3        4   24   M  technician    43537
4        5   33   F       other    15213

So, we have 943 users in the dataset and each user has 5 features, i.e. user_ID, age, sex, occupation and zip_code

In [7]:
# rating file
print(ratings.shape)
ratings.head()

(100000, 4)


user_id  movie_id  rating  unix_timestamp
0      196       242       3       881250949
1      186       302       3       891717742
2       22       377       1       878887116
3      244        51       2       880606923
4      166       346       1       886397596

So, We have 100k ratings for different user and movie combinations.

In [8]:
# item files
print(items.shape)
items.head(20)

(1682, 24)


movie id                                        movie title release date  \
0          1                                   Toy Story (1995)  01-Jan-1995   
1          2                                   GoldenEye (1995)  01-Jan-1995   
2          3                                  Four Rooms (1995)  01-Jan-1995   
3          4                                  Get Shorty (1995)  01-Jan-1995   
4          5                                     Copycat (1995)  01-Jan-1995   
5          6  Shanghai Triad (Yao a yao yao dao waipo qiao) ...  01-Jan-1995   
6          7                              Twelve Monkeys (1995)  01-Jan-1995   
7          8                                        Babe (1995)  01-Jan-1995   
8          9                            Dead Man Walking (1995)  01-Jan-1995   
9         10                                 Richard III (1995)  22-Jan-1996   
10        11                               Seven (Se7en) (1995)  01-Jan-1995   
11        12                         Usual Suspects, The (1995)  14-Aug-1995   
12        13                            Mighty Aphrodite (1995)  30-Oct-1995   
13        14                                 Postino, Il (1994)  01-Jan-1994   
14        15                          Mr. Holland's Opus (1995)  29-Jan-1996   
15        16                 French Twist (Gazon maudit) (1995)  01-Jan-1995   
16        17                         From Dusk Till Dawn (1996)  05-Feb-1996   
17        18                          White Balloon, The (1995)  01-Jan-1995   
18        19                              Antonia's Line (1995)  01-Jan-1995   
19        20                          Angels and Insects (1995)  01-Jan-1995   

    video release date                                           IMDb URL  \
0                  NaN  http://us.imdb.com/M/title-exact?Toy%20Story%2...   
1                  NaN  http://us.imdb.com/M/title-exact?GoldenEye%20(...   
2                  NaN  http://us.imdb.com/M/title-exact?Four%20Rooms%...   
3                  NaN  http://us.imdb.com/M/title-exact?Get%20Shorty%...   
4                  NaN  http://us.imdb.com/M/title-exact?Copycat%20(1995)   
5                  NaN  http://us.imdb.com/Title?Yao+a+yao+yao+dao+wai...   
6                  NaN  http://us.imdb.com/M/title-exact?Twelve%20Monk...   
7                  NaN     http://us.imdb.com/M/title-exact?Babe%20(1995)   
8                  NaN  http://us.imdb.com/M/title-exact?Dead%20Man%20...   
9                  NaN  http://us.imdb.com/M/title-exact?Richard%20III...   
10                 NaN    http://us.imdb.com/M/title-exact?Se7en%20(1995)   
11                 NaN  http://us.imdb.com/M/title-exact?Usual%20Suspe...   
12                 NaN  http://us.imdb.com/M/title-exact?Mighty%20Aphr...   
13                 NaN  http://us.imdb.com/M/title-exact?Postino,%20Il...   
14                 NaN  http://us.imdb.com/M/title-exact?Mr.%20Holland...   
15                 NaN  http://us.imdb.com/M/title-exact?Gazon%20maudi...   
16                 NaN  http://us.imdb.com/M/title-exact?From%20Dusk%2...   
17                 NaN  http://us.imdb.com/M/title-exact?Badkonake%20S...   
18                 NaN  http://us.imdb.com/M/title-exact?Antonia%20(1995)   
19                 NaN  http://us.imdb.com/M/title-exact?Angels%20and%...   

    unknown  Action  Adventure  Animation  Children's  ...  Fantasy  \
0         0       0          0          1           1  ...        0   
1         0       1          1          0           0  ...        0   
2         0       0          0          0           0  ...        0   
3         0       1          0          0           0  ...        0   
4         0       0          0          0           0  ...        0   
5         0       0          0          0           0  ...        0   
6         0       0          0          0           0  ...        0   
7         0       0          0          0           1  ...        0   
8         0       0          0          0           0  ...        

This dataset contains attributes of 1682 movies. There are 24 columns out of which last 19 columns specify the genre of a particular movie. These are binary columns, i.e., a value of 1 denotes that the movie belongs to that genre, and 0 otherwise.

In [9]:
r_cols = ['user_id', 'movie_id', 'rating', 'unix_timestamp']
ratings_train = pd.read_csv('ml-100k/ua.base', sep='\t', names=r_cols, encoding='latin-1')
ratings_test = pd.read_csv('ml-100k/ua.test', sep='\t', names=r_cols, encoding='latin-1')
ratings_train.shape, ratings_test.shape

((90570, 4), (9430, 4))

The dataset has already been divided into train and test by GroupLens where the test data has 10 ratings for each user, i.e. 9,430 rows in total. We will read both these files into our Python environment.

### Building collaborative filtering model

We will recommend movies based on user-user similarity and item-item similarity. For that, first we need to calculate the number of unique users and movies.

In [10]:
n_users = ratings.user_id.unique().shape[0]
n_items = ratings.movie_id.unique().shape[0]

Now, we need to create a user-item matrix which can be used to calculate the similarity between users and items.

In [11]:
data_matrix = np.zeros((n_users, n_items))
for line in ratings.itertuples():
    data_matrix[line[1]-1, line[2]-1] = line[3]

Now, we will calculate the similarity. We can use the pairwise_distance function from sklearn to calculate the cosine similarity.

In [12]:
from sklearn.metrics.pairwise import pairwise_distances 
user_similarity = pairwise_distances(data_matrix, metric='cosine')
item_similarity = pairwise_distances(data_matrix.T, metric='cosine')

This gives us the item-item and user-user similarity in an array form. The next step is to make predictions based on these similarities. Let’s define a function to do just that.

In [13]:
def predict(ratings, similarity, type='user'):
    if type == 'user':
        mean_user_rating = ratings.mean(axis=1)
        #We use np.newaxis so that mean_user_rating has same format as ratings
        ratings_diff = (ratings - mean_user_rating[:, np.newaxis])
        pred = mean_user_rating[:, np.newaxis] + similarity.dot(ratings_diff) / np.array([np.abs(similarity).sum(axis=1)]).T
    elif type == 'item':
        pred = ratings.dot(similarity) / np.array([np.abs(similarity).sum(axis=1)])
    return pred

Finally, we will make predictions based on user similarity and item similarity.

In [14]:
user_prediction = predict(data_matrix, user_similarity, type='user')
item_prediction = predict(data_matrix, item_similarity, type='item')

### Building a simple popularity and collaborative filtering model using Turicreate